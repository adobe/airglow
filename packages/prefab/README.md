# Prefab

A framework for writing reusable reducers and selectors in a [Redux](redux.js.org) application.

## Introduction

In a [Redux](redux.js.org) applications, we hold the entire application state in a single _store_. We extract data from the _store_ using __selectors__. We add new data into the store using __reducers__. Prefab is a framework allowing you to create __selectors__ and __reducers__ that your components can reuse.

For example, let's say we have a map component:

![Sample Map](https://git.corp.adobe.com/dms-mobile/dipper/blob/master/docs/assets/sample.map.png?raw=true)

Every time our map component is used, there are certain things we know we'll have to hook up. We need to know the current coordinates of the map. We need to handle zooming in and out. We need to know whether to show a street map or satelite map. Using __Prefab__, we can consolidate all of the common map logic into a __Prefab Construct__.

## Initializing Prefab

### With Airglow

Initializing __Prefab__ in Airglow is simple:

```
import Airglow, { BootstrapWrapper } from '@airglow/core';
import prefabPlugin from '@airglow/prefab-plugin';

export const App = () => (
  <Airglow plugins={[prefabPlugin()]}>
    <BootstrapWrapper
      config={{ prefab: prefabConfig }}>
        ...application
    </BootstrapWrapper>
  </Airglow>
);
```

### Standalone

```
import { createStore, compose } from 'redux';
import { reducer, BOOTSTRAP_PREFAB } from '@airglow/prefab';

const store = createStore(
  combineReducers({
    prefab: reducer
  })
);

store.dispatch({ type: BOOTSTRAP_PREFAB, payload: prefabConfig });
```

Reguardless, __Prefab__ will add a `prefab` section to the store where it will hold all it's information.


## Constructs

__Constructs__ are a group of rules and settings that the __Prefab Framework__ is able to use. Continuing with our map sample, look at the following construct:

```
const mapConstruct = ({
  defaultZoom = 1
}) => ({
  defaultZoom
});

const prefabConfig = {
  bigMap: mapConstruct(2),
  smallMap: mapConstruct(6)
};
```

When applying this config to __Prefab__, __Prefab__ our _Redux store_ will look like:

```
prefab: {
  bigMap: {
    construct: {
      name: 'bigMap',
      defaultZoom: 2
    },
    store: {
    }
  },
  smallMap: {
    construct: {
      name: 'smallMap',
      defaultZoom: 6
    },
    store: {
    }
  }
}
```

All of the original configuration gets stored in the _construct_ section. All changes to state will be stored in the _store_ section.

## Selectors

As you no doubt noticed, part of creating the _mapConstruct_ was defining a __selector__. This is how we are going to pull data out of the store. Let's take a look at a sample _mapSelector_:

```
const mapSelector = ({ state }) => ({
  zoom: state.store.zoom || state.construct.defaultZoom
});
```

This is a really basic __selector__. We are passing in the current zoom value from the store. If there isn't one, we are passing in the defaultZoom.

## Event Handling

Next, it's likely we'll want to handle events in our __Ubra Construct__ as well:

```
const mapSelector = ({ state, dispatch }) => ({
  zoom: state.store.zoom || state.construct.defaultZoom,
  onZoomIn: dispatch({ type: ZOOM_IN }),
  onZoomOut: dispatch({ type: ZOOM_OUT })
});
```

## Reducers

We'll want to automatically handle those _zoom_ events:

```
const reduce = (state, action) => {
  const zoom = state.store.zoom || state.construct.defaultZoom;
  if (action.type === ZOOM_IN) {
    return { ...state, store: { ...store, zoom: zoom + 1 } };
  }
  if (action.type === ZOOM_OUT) {
    return { ...state, store: { ...store, zoom: zoom + 1 } };
  }
}
```

Here we are handling the _Redux actions_ and updating the store accordingly.

We need to register this reducer inside our __construct__:

```
const mapConstruct = ({
  defaultZoom = 1
}) => ({
  defaultZoom,
  reduce,
  selector: mapSelector
});
```

## Using our Construct

So we now have a construct that provides the current zoom level and a handler for updating that zoom level. Let's show how easy this is to use in our code:

```
import connect from '@airglow/deep-connect';
import { selector } from '@airglow/prefab';

const MapApp = (smallMap, largeMap) => {
	<div>
      <Map {...smallMap} />
      <Map {...largeMap} />
	</div>
}

const mapForm = selector({
  smallMap: 'smallMap',
  largeMap: 'largeMap'
});

export default connect({ mapForm }, Component);
```
