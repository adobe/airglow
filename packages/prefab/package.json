{
  "name": "@airglow/prefab",
  "version": "0.1.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/adobe/airglow.git",
    "directory": "packages/prefab"
  },
  "description": "Prefabicated Selectors and Reducers",
  "author": "Adobe Systems",
  "license": "Apache-2.0",
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "registry": "https://artifactory.corp.adobe.com/artifactory/api/npm/npm-nubula-release"
  },
  "dependencies": {
    "@airglow/reducers": "^0.1.0",
    "ramda": "^0.26.1",
    "reselect": "^4.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.4.0"
  },
  "scripts": {
    "prebuild": "rm -rf dist && mkdir dist",
    "build": "babel src --out-dir dist --root-mode upward"
  },
  "main": "dist/index.js"
}
