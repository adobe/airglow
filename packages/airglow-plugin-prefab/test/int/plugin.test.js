/*
Copyright 2019 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

import React from 'react';
import { renderAirglow } from '@airglow/test-utils';
import ReactStore, { connect } from '@airglow/store-react';
import prefab from '@airglow/prefab';

import plugin from '../../src/index';

let tree;

const prefabs = prefab({ color: { type: 'value' } });

const mapState = state => ({ color: prefabs.color.value(state) });

const mapHandlers = dispatch => ({
  onColor: v => dispatch(prefabs.color.changeAction(v))
});

const TestComponent = ({ color, onColor }) => (
  <input value={color || ''} onChange={e => onColor(e.target.value)} />
);

const TestConnected = connect(mapState, mapHandlers)(TestComponent);
const configuredPlugin = plugin();


describe('AirglowPrefabIntegrationTest', () => {
  beforeEach(() => {
    tree = renderAirglow(
      <TestConnected />,
      {
        store: ReactStore,
        plugins: [configuredPlugin],
        bootstrap: { prefab: prefabs }
      }
    );
  });
  afterEach(() => {
  });

  it('should render with default values', function () {
    expect(tree.debug()).toMatchSnapshot();
  });

  it('should call the reducer when the text changes', function (done) {
    tree.find('input').simulate('change', { target: { value: 'Purple!' } });

    setTimeout(() => {
      expect(tree.debug()).toMatchSnapshot();
      done();
    }, 30);
  });
});
