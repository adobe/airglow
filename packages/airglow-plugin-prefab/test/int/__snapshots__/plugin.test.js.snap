// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AirglowPrefabIntegrationTest should call the reducer when the text changes 1`] = `
"<App>
  <Airglow store={[Function]} plugins={{...}}>
    <StoreProvider store={{...}}>
      <AirglowWrapper config={{...}}>
        <Component>
          <TestComponent color=\\"Purple!\\" onColor={[Function: onColor]}>
            <input value=\\"Purple!\\" onChange={[Function: onChange]} />
          </TestComponent>
        </Component>
      </AirglowWrapper>
    </StoreProvider>
  </Airglow>
</App>"
`;

exports[`AirglowPrefabIntegrationTest should render with default values 1`] = `
"<App>
  <Airglow store={[Function]} plugins={{...}}>
    <StoreProvider store={{...}}>
      <AirglowWrapper config={{...}}>
        <Component>
          <TestComponent color=\\"\\" onColor={[Function: onColor]}>
            <input value=\\"\\" onChange={[Function: onChange]} />
          </TestComponent>
        </Component>
      </AirglowWrapper>
    </StoreProvider>
  </Airglow>
</App>"
`;
